File: app/src/test/java/de/westnordost/streetcomplete/data/meta/OsmAreasTest.java
package de.westnordost.streetcomplete.data.meta;

import org.junit.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import de.westnordost.osmapi.map.data.OsmRelation;
import de.westnordost.osmapi.map.data.OsmWay;
import de.westnordost.osmapi.map.data.Way;

import static org.junit.Assert.*;

public class OsmAreasTest
{
	@Test public void relation()
	{
		assertFalse(OsmAreas.isArea(new OsmRelation(0,0, null, null)));

		Map<String, String> tags = new HashMap<>();
		tags.put("type","multipolygon");

		assertTrue(OsmAreas.isArea(new OsmRelation(0, 0, null, tags)));
	}

	@Test public void wayNoTags()
	{
		assertFalse(OsmAreas.isArea(createWay(false, null)));
		assertFalse(OsmAreas.isArea(createWay(true, null)));
	}

	@Test public void waySimple()
	{
		Map<String, String> tags = new HashMap<>();
		tags.put("area","yes");

		assertFalse(OsmAreas.isArea(createWay(false, tags)));
		assertTrue(OsmAreas.isArea(createWay(true, tags)));
	}

	@Test public void wayRuleException()
	{
		Map<String, String> tags = new HashMap<>();
		tags.put("railway","something");

		assertFalse(OsmAreas.isArea(createWay(true, tags)));

		tags.put("railway","station");

		assertTrue(OsmAreas.isArea(createWay(true, tags)));
	}

	@Test public void wayRuleRegex()
	{
		Map<String, String> tags = new HashMap<>();
		tags.put("waterway","duck");

		assertFalse(OsmAreas.isArea(createWay(true, tags)));

		tags.put("waterway","dock");

		assertTrue(OsmAreas.isArea(createWay(true, tags)));
	}

	private Way createWay(boolean ring, Map<String,String> tags)
	{
		List<Long> nodes = ring ? Arrays.asList(0L,1L,0L) : Arrays.asList(0L,1L);
		return new OsmWay(0, 0, nodes, tags );
	}
}

File: app/src/test/java/de/westnordost/streetcomplete/data/meta/OsmAreasTest.kt
package de.westnordost.streetcomplete.data.meta

import org.junit.Test

import de.westnordost.osmapi.map.data.OsmRelation
import de.westnordost.osmapi.map.data.OsmWay
import de.westnordost.osmapi.map.data.Way
import de.westnordost.streetcomplete.data.meta.OsmAreas.isArea

import org.junit.Assert.*

class OsmAreasTest {
    @Test fun `relation with no tags is no area`() {
        assertFalse(isArea(OsmRelation(0, 0, null, null)))
    }

    @Test fun `multipolygon relation is an area`() {
        assertTrue(isArea(OsmRelation(0, 0, null, mapOf("type" to "multipolygon"))))
    }

    @Test fun `way with no tags is no area`() {
        assertFalse(isArea(createWay(false, null)))
        assertFalse(isArea(createWay(true, null)))
    }

    @Test fun `simple way with area=yes tag is no area`() {
        assertFalse(isArea(createWay(false, mapOf("area" to "yes"))))
    }

    @Test fun `closed way with area=yes tag is an area`() {
        assertTrue(isArea(createWay(true, mapOf("area" to "yes"))))
    }

    @Test fun `closed way with specific value of a key that is usually no area is an area`() {
        assertFalse(isArea(createWay(true, mapOf("railway" to "something"))))
        assertTrue(isArea(createWay(true, mapOf("railway" to "station"))))
    }

    @Test fun `closed way with a certain tag value is an area`() {
        assertFalse(isArea(createWay(true, mapOf("waterway" to "duck"))))
        assertTrue(isArea(createWay(true, mapOf("waterway" to "dock"))))
    }

    private fun createWay(ring: Boolean, tags: Map<String, String>?): Way {
        val nodes = if (ring) listOf(0L, 1L, 0L) else listOf(0L, 1L)
        return OsmWay(0, 0, nodes, tags)
    }
}
